openmls/TEST-DESIGN.md

```diff
@@ -0,0 +1,176 @@
+      Rule of thumb could be `1 method = 1 mod` but it can be otherwise.
+    * Imports SHOULD live in the top `tests` mod to avoid duplicates. Nested mods SHOULD not have anything other
+      than `use super::*;`
+    * Within a nested mod you can avoid repeating the mod name e.g.
+      ~~`method_a_should_bla_when_blabla`~~ -> `should_bla_when_blabla`
+    ```rust
+    impl MyStruct {
+        pub fn method_a() {}
+        // snip
+        fn method_z() {}
+    }
+    
+    // No `#[cfg(features = "test-utils")]`
+    
+    #[test]
+    mod tests {
+        use super::*;
+        use crate::prelude::*; // if required
+    
+        mod a {
+            use super::*; // <- no other imports here. 
+            // If one is required, add it to tests mod. This avoids duplicates.
+    
+            #[test]
+            fn should_bla_when_blabla_1() {}
+    
+            // snip
+            #[test]
+            fn should_bla_when_blabla_n() {}
+        }
+    
+        // snip
+    
+        mod z {
+            use super::*;
+    
+            #[test]
+            fn should_bla_when_blabla_1() {}
+    
+            // snip
+            #[test]
+            fn should_bla_when_blabla_n() {}
+        }
+    }
+    ```
+
+12. Crate layout
+    * unit tests in `src` in the file they relate to
+    * a `test_utils` mod SHOULD live in each top level mod. It SHOULD be declared with `#[cfg(test)]` in parent `mod.rs`
+    * Those test utils SHOULD contain all the reusable helpers to write unit tests in all the other inner files.
+    * Those test utils are public and can be used by other top level mods e.g. `z` can use `a`'s utils
+    ```text
+    src
+    ├── a
+    │ ├── mod.rs <- unit tests here
+    │ └── test_utils.rs <- public helpers for unit tests #[cfg(test)]
+    │ -- snip --
+    ├── z
+    │ ├── mod.rs
+    │ └── test_utils.rs
+    tests <- integration tests with public api only
+    ├── a.rs
+    │ -- snip --
+    └── z.rs
+    ```
+
+13. Integration tests live in `./tests` folder. 
+    1. There SHOULD be one mod per source mod. 
+    2. Integration tests SHOULD only use the public API.
+    3. They could be used as examples in the book
+
+## Bonus
+
+* We could use mutation testing to "test out tests". A good crate available for doing that
+  is [mutagen](https://github.com/llogiq/mutagen). After annotating all the methods in the [`tree` mod](src/tree) it
+  gives the [following result](https://github.com/beltram/openmls/blob/mutation/openmls/MUTATION.txt) with ~= 1/3 of
+  mutants surviving i.e. not rightfully tested
+* If WASM is a target we could already use [wasm-bindgen-test](https://crates.io/crates/wasm-bindgen-test) macros on all
+  tests methods (no async migration required at that point, only a few things to fix: rayon, time, rng)
+* Use [nextest](https://crates.io/crates/cargo-nextest) to execute test in the CI because it is faster (19s vs 30s for
+  cargo test)
```

Comment:
I always use that locally, we could consider it for the CI as well if there are no downsides.

That being said, the overall speedup on the CI should only be marginal, since building is the slowest part.